### PR TITLE
Add support for Redis Cache

### PR DESCRIPTION
**Pull request by yadavsahil197**
Adding support for Redis as an cache endpoint. The Redis cache endpoint can be used for both generations and transformations by passing an instance of the RedisCache to the Labeling Agent.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -28,6 +28,7 @@ dependencies = [
     "scikit-learn >= 1.0.0",
     "tenacity >= 8.2.2",
     "SQLAlchemy >= 2.0.19",
+    "redis >= 3.5.3",
```
Comment (nihit) [inline pyproject.toml:31]: let's make this an optional (extension) dependency ?

**Review by nihit — APPROVED** [review]
lgtm otherwise

File: src/autolabel/labeler.py [hunk 2]
```diff
@@ -53,15 +57,28 @@ class LabelingAgent:
...
-        self.generation_cache = SQLAlchemyGenerationCache() if cache else None
-        self.transform_cache = SQLAlchemyTransformCache() if cache else None
+        self.generation_cache = generation_cache
+        self.transform_cache = transform_cache
+        if cache:
+            logger.warning(
+                f"cache parameter is deprecated and will be removed soon. Please use generation_cache and transform_cache instead."
+            )
+            self.generation_cache = (
+                generation_cache if generation_cache else SQLAlchemyGenerationCache()
```
Comment (rajasbansal) [inline src/autolabel/labeler.py:76]: What would be the way to disable caches if cache parameter is going to be depracated?

File: src/autolabel/schema.py [hunk 3]
```diff
@@ -209,6 +210,24 @@ class GenerationCacheEntry(BaseModel):
...
+        """
+        Generates a unique ID for the given generation cache configuration
+        """
+        return calculate_md5([self.model_name, self.model_params, self.prompt])
+
+    def get_serialized_output(self) -> str:
+        """
+        Returns the serialized cache entry output
+        """
+        return json.dumps([gen.dict() for gen in self.generations])
```
Comment (rajasbansal) [inline src/autolabel/schema.py:223]: For things like bytes objects which come as a result of transforms, json dumping fails. That is why we switched to serializing using pickle.

File: src/autolabel/labeler.py (hunk 2 above)
Comment (yadavsahil197) [reply src/autolabel/labeler.py:76]: The user can pass the cache objects directly to generation_cache and transform_cache to use the cache if needed.

File: src/autolabel/schema.py (hunk 3 above)
Comment (yadavsahil197) [reply src/autolabel/schema.py:223]: Makes sense, I can move to pickle in a follow up PR

**Pull request merged**